src/plots/cartesian/index.js

```diff
@@ -25,61 +25,74 @@ exports.attrRegex = constants.attrRegex;
 
 exports.attributes = require('./attributes');
 
-exports.plot = function(gd) {
+exports.plot = function(gd, traces) {
+    var cdSubplot, cd, trace, i, j, k, isFullReplot;
+
     var fullLayout = gd._fullLayout,
         subplots = Plots.getSubplotIds(fullLayout, 'cartesian'),
         calcdata = gd.calcdata,
         modules = fullLayout._modules;
 
-    function getCdSubplot(calcdata, subplot) {
-        var cdSubplot = [];
-
-        for(var i = 0; i < calcdata.length; i++) {
-            var cd = calcdata[i];
-            var trace = cd[0].trace;
-
-            if(trace.xaxis + trace.yaxis === subplot) {
-                cdSubplot.push(cd);
-            }
+    if(!Array.isArray(traces)) {
+      // If traces is not provided, then it's a complete replot and missing
+      // traces are removed
+        isFullReplot = true;
+        traces = [];
+        for(i = 0; i < calcdata.length; i++) {
+            traces.push(i);
         }
-
-        return cdSubplot;
+    } else {
+      // If traces are explicitly specified, then it's a partial replot and
+      // traces are not removed.
+        isFullReplot = false;
```

Comment:
Disregard previous comment. Reason: was avoiding doing the keyed data join in `Cartesian.plot`. Since it's `Scatter.plot` that does it, it needs to know whether to remove un-updated traces.